python/lsst/daf/butler/queries/_dataset_query_results.py

```diff
@@ -0,0 +1,236 @@
+        self._spec = spec
+
+    def __iter__(self) -> Iterator[DatasetRef]:
+        page = self._driver.execute(self._spec, self._tree)
+        yield from page.rows
+        while page.next_key is not None:
+            page = self._driver.fetch_next_page(self._spec, page.next_key)
+            yield from page.rows
+
+    @property
+    def dataset_type(self) -> DatasetType:
+        # Docstring inherited.
+        return DatasetType(self._spec.dataset_type_name, self._spec.dimensions, self._spec.storage_class_name)
+
+    @property
+    def data_ids(self) -> DataCoordinateQueryResults:
+        # Docstring inherited.
+        from ._data_coordinate_query_results import DataCoordinateQueryResults
+
+        return DataCoordinateQueryResults(
+            self._driver,
+            tree=self._tree,
+            spec=DataCoordinateResultSpec.model_construct(
+                dimensions=self.dataset_type.dimensions.as_group(),
+                include_dimension_records=self._spec.include_dimension_records,
+            ),
+        )
+
+    @property
+    def has_dimension_records(self) -> bool:
+        # Docstring inherited.
+        return self._spec.include_dimension_records
+
+    def with_dimension_records(self) -> SingleTypeDatasetQueryResults:
+        # Docstring inherited.
+        if self.has_dimension_records:
+            return self
+        return self._copy(tree=self._tree, include_dimension_records=True)
+
+    def by_dataset_type(self) -> Iterator[SingleTypeDatasetQueryResults]:
+        # Docstring inherited.
+        return iter((self,))
+
+    def count(self, *, exact: bool = True, discard: bool = False) -> int:
+        # Docstring inherited.
+        return self._driver.count(self._tree, self._spec, exact=exact, discard=discard)
+
+    def _copy(self, tree: QueryTree, **kwargs: Any) -> SingleTypeDatasetQueryResults:
+        # Docstring inherited.
+        return SingleTypeDatasetQueryResults(self._driver, tree, self._spec.model_copy(update=kwargs))
+
+    def _get_datasets(self) -> frozenset[str]:
+        # Docstring inherited.
+        return frozenset({self.dataset_type.name})
+
+
+@final
+class ChainedDatasetQueryResults(DatasetQueryResults):
+    """A query for `DatasetRef` results with multiple dataset types.
+
+    Parameters
+    ----------
+    by_dataset_type : `tuple` [ `SingleTypeDatasetQueryResults` ]
+        Tuple of single-dataset-type query result objects to combine.
+
+    Notes
+    -----
+    This class should never be constructed directly by users; use
+    `Query.datasets` instead.
+    """
+
+    def __init__(self, by_dataset_type: tuple[SingleTypeDatasetQueryResults, ...]):
+        self._by_dataset_type = by_dataset_type
+
+    def __iter__(self) -> Iterator[DatasetRef]:
+        return itertools.chain.from_iterable(self._by_dataset_type)
+
+    def by_dataset_type(self) -> Iterator[SingleTypeDatasetQueryResults]:
+        # Docstring inherited.
+        return iter(self._by_dataset_type)
```

Comment:
Cool, that seems like it will buy us a lot of simplicity.  Seems like it should be easy enough to add back a multi/batch version later if it turns out we want it.

Reply:
Done on DM-43146.